Play-2.4 version

sovaalexandr opened the pull request:
DI - oriented and Play! Framework 2.4 ready version of Geo-location module

megazord commented on .travis.yml:
```diff
@@ -1,12 +1,10 @@
 language: scala
 scala:
-   - 2.10.4
```
We should not drop support for Scala 2.10, since Play 2.4 stills supports it.

sovaalexandr replied:
Is it really so? As I understood from https://playframework.com/documentation/2.4.x/Migration24#Java-8-support new projects should set scalaVersion to 2.11.6+

sovaalexandr replied:
Returned back

megazord commented on project-code/app/com/edulify/modules/geolocation/GeolocationFactory.java:
```diff
@@ -0,0 +1,34 @@
+package com.edulify.modules.geolocation;
+
+/**
+ * Created by sowhat
```
Could you please remove these headers?

sovaalexandr replied:
Ok. Sorry, default Idea setting

sovaalexandr replied:
Removed

megazord commented on project-code/app/com/edulify/modules/geolocation/GeolocationFactory.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Created by sowhat
+ */
+public class GeolocationFactory {
```
I still don't understand why we need this.

sovaalexandr replied:
As I commented before - it's to have an ablity to make isolated unit tests and have a control over instantiation of Geolocation object (for example in tests)
Also this decomposition is a way to follow the Single Responsibility principle

megazord replied:
But, since `Geolocation` is a simple POJO, do we really need a factory here? It looks like a overkill to me.

sovaalexandr replied:
Will do instantiation without factory

megazord commented on project-code/app/com/edulify/modules/geolocation/GeolocationModule.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public class GeolocationModule extends AbstractModule {
+
+    private final Environment environment;
```
`Environment` variable is not used here. Could it be removed?

sovaalexandr replied:
Can try. I was guided by https://playframework.com/documentation/2.4.x/JavaDependencyInjection#Configurable-bindings while implementing this

sovaalexandr replied:
Yeah, as I expected - You can't
The Module instantiation code is at  `play.api.inject.Modules:118` and it requires two-argument, or none-argument constructor that is not our case. Or throws `PlayException("No valid constructors"....`

megazord commented on project-code/app/com/edulify/modules/geolocation/GeolocationPlugin.java:
```diff
@@ -1,39 +0,0 @@
-import com.edulify.modules.geolocation.providers.FreegeoipProvider;
-import play.Application;
-
-public class GeolocationPlugin extends play.Plugin {
```
I don't think we should remove the plugin class. We could just deprecated it, like Play did, and let users know that it will be remove soon.

sovaalexandr replied:
Yeah, but it'll lead to situation when we'll have two similar component infrastructures one at plugin and one at DI container + the plugin infrastructure will be singleton. Should I really do It?

megazord replied:
It should be documented with a warning that users cannot use both DI Container + Plugin infrastructure together. Or, if it possible, we should avoid the plugin to start if we already have a DI component.

But, for this PR, you can just add the plugin back and after a merge we can work together to figure out how DI container and plugins will live harmoniously.

sovaalexandr replied:
Ok, I'll add it back with standalone infrastructure

megazord commented on project-code/app/com/edulify/modules/geolocation/GeolocationService.java:
```diff
@@ -1,14 +1,27 @@
+  public Promise<Geolocation> getGeolocation(String ip) {
+    if (null == ip) return pure(null);
+    if (ip.isEmpty()) return pure(null);
+    Geolocation geolocation = cache.get(ip);
```
Maybe `GeolocationCache` should expose a `getOrElse` method, like Play cache does, and then do something like:

```java
cache.getOrElse(ip, ipToGeolocate -> { ... });
```

sovaalexandr replied:
Good solution. Let's try to make it.

sovaalexandr replied:
Unfortunately can't see a way to do it because of:

 - Play's Cache stores retrieved object at cache without asking do we need it - too bad
 - Even if code will look some like this
```java
//.... class GeolocationCache ....
  public Geolocation getOrElse(String ip, Callable<Geolocation> block) throws Exception {
    return cacheOn ? Cache.getOrElse(key(ip), block, 0) : block.call();
  }
```
we should use Callable that returns `Geolocation` what means that we can't use `GeolocationProvider` instances because those instances returns `Promise<Geolocation>` and we can't get plain `Geolocation` object (without blocking, that is definitely no that we want)

So I'd prefer to leave it as is. But may be You have a vision/solution for this case?

megazord replied:
Lets work to integrate this PR and then we can think about this kind of optimization.

megazord commented:
@sovaalexandr thank you so much for this PR. I don't have a problem with it being created to master, since there are no differences between master and play-2.4.0 branches.

Please, consider my comments more as questions. I really appreciate your help and time to move this module forward.

sovaalexandr commented:
@megazord thank you having time to review my code. I'll make changes to this branch ASAP

sovaalexandr commented:
@megazord 
I had just finished my changes on returning back plugin and corrected documentation
But right now provider code seems uglier to me. For instance I'd prefer to use constructor injection rather than setter-injection and so on.
But in considerations of backwards-compatibility I tried to make plugin and provider instantiation as flexible as possible.
I also removed `GeolocationFactory` and processed each your feedback. But I still want to make code of this component more strict.

megazord commented:
@sovaalexandr this PR was merged manually into branch `play-2.4.0`. There are some points that I want to change before doing a release (or maybe a milestone release). If you decide to submit more pull requests, please submit them against branch `play-2.4.0`.

Thank you so much!

sovaalexandr commented:
@megazord Welcome)
You can also close issue #7
